README.md
# Multi-Service RTMP Broadcaster

The goal of this project is to create a Docker-deployed service that will allow you to easily broadcast a live stream to multiple services (YouTube, Twitch, Facebook, etc) at the same time. This stream rebroadcaster is designed to be used by a single source / user.

## Usage
The instructions here assume you are running on Linux. With some modification of the commands, you can make this Docker build work on Windows and Mac OS too.

The first step is to build the docker file. After you clone this repository, `cd` into and and issue:

```sh
docker build -t multistreaming-server ./multistreaming-server/
```

Once built, start the docker image on a host that has sufficient bandwidth to handle all the rebroadcasting you will do:

```sh
docker run -it -p 80:80 -p 1935:1935 \
  --env MULTISTREAMING_PASSWORD=__made_up_password__ \
  -v /path/to/my-rtmp-config.json:/rtmp-configuration.json \
  multistreaming-server:latest
```

If this is a host than where you built the docker image, you will need to push the docker image to that host (or build it there). Alternatively, you could use the DockerHub build of this image by pulling and using the `kamprath/multistreaming-server:latest` [Docker image](https://hub.docker.com/repository/docker/kamprath/multistreaming-server). Also, if you plan on doing any transcoding when rebroadcasting a stream, you need to ensure that your docker host's CPU is sufficient for the job. It is recommend that the host CPU has at least four cores for each distinct transcoding the multi-streaming server will do. If the host CPU is not sufficient, bit rates on the transcoded streams will suffer.

Note that some environment variables should be set when running the Docker image:

* `MULTISTREAMING_PASSWORD` _(REQUIRED)_ - This is a password you define and will be used by your steaming software. This is a marginally secure way to prevent other people from pushing to your stream.
* `CONFIG_DISABLE_RECORD` _(OPTIONAL)_ - If this env var is set to a non-empty value, the stream recording feature will be disabled. 
* `CONFIG_NGINX_DEBUG` _(OPTINAL)_ - If this env var is set to a non-empty value, it will enable the nginx debug logging to stderr. This is useful for debugging.
* `CONFIG_FFMPEG_LOG` _(OPTIONAL)_ - If this env var is set to a non-empty value, each ffmpeg process will start logging their output to `/tmp/ffmpeg-$APP.log` in the container. This is useful for debugging issues related to encoding.
* `CONFIG_FFMPEG_MAX_MUXING_QUEUE_SIZE` _(OPTIONAL)_ - This env var must be a numerical value, it will be used as the global default for `-max_muxing_queue_size` argument for the ffmpeg processes. You may need to set this to a higher value (i.e: `4096`) if you run in to this [bug.](https://trac.ffmpeg.org/ticket/6375)

You must also create and JSON file with the RTMP rebroadcasting configuration you desire. This file should get mapped from your local file system to the `/rtmp-configuration.json` file path within the Docker container. The JSON file has the following elements:

* `endpoint`- This is the name of the RTMP ingest endpoint that the source stream will be pushed to. Defaults to `live` if not specified.
* `transcodeProfiles` - This is a JSON object with keys indicating a unique identifier for a transcode profile, and the values specifying the supported transcode options.
    * `pixels` - The pixel dimension that the stream should be transcoded to. Formatted like "1920x1080". If not specified, defaults to "1280x720".
    * `videoBitRate` - The video bit rate that should be used when sending the stream to this destination. Should be a number followed by "k" or "m" for kilo- and mega- bits-per-second. If not specified defaults to "4500k".
    * `videoKeyFrameSecs` - The number of seconds between key frames in the transcoded stream. If not specified, defaults to 2.
    * `videoFrameRate` - The frames per second which the video stream will be re-encoded with, defaults to "30".
    * `audioBitRate` - The bit rate that should be used for the transcoded audio signal. Should be a number followed by "k" or "m" for kilo- and mega- bits-per-second. If not specified, defaults to "160k". If neither `audioBitRate` or `audioSampleRate` are specified, then the audio signal is simply copied from source with no alteration.
    * `audioSampleRate` - The sampling rate to be used for the transcoded audio signal. Should be an integer indicating the sampling Hertz. If not specified, defaults to `48000`. If neither `audioBitRate` or `audioSampleRate` are specified, then the audio signal is simply copied from source with no alteration.
    * `maxMuxingQueueSize` - Individual override for the `CONFIG_FFMPEG_MAX_MUXING_QUEUE_SIZE` env var. See the env var section for the usage.
* `rebroadcastList`- _Required_ Contains a list of JSON objects that each configure a distinct RTMP destination that the stream pushed to the ingest endpoint will be rebroadcasted to. At least one destination should be configured. There is no specific limit on the number of destinations except for the hardware limitations of your host. Each destination is configured with the following JSON elements:
  * `name` - _Required_ A distinct label for this destination. Can be any alphanumeric string with no white space. Must be distinct from all the other destination names in this list.
  * `platform` - _Required_ The platform that this specific rebroadcast stream should be pushed to. The default RTMP destinations will be used for each platform. Supported platforms values are: `youtube`, `facebook`, `twitch`, `instagram`, `periscope`, `microsoft-stream`, `mixcloud`, `dlive` and `custom`. Note that specifying `microsoft-stream` or `custom` will cause the `streamKey` element to be ignored if present and instead use the `fullRTMPURL`
  * `regionCode` - If `periscope` is specified as the platform for this destination, this is the two letter region code that is part of the Periscope server URL. If undefined, it will default to `ca` (the "US West" region)
  * `streamKey` - This is the stream key that identifies the unique stream on the specified platform. This value is provided by the platform. This element must be provided for all `platform` types except for `custom` and `microsoft-stream`.
  * `fullRTMPURL` - If `custom` or `microsoft-stream` is specified in the `platform`, the URL specified in this element is used for the forming destination URL. This should include the `rtmp://` prefix. For the `microsoft-stream` platform, this is the full URL provided in their stream set up. This element is ignored for all other platform types.
  * `transcode` - If present, the stream will be transcoded before rebroadcasting it to this list item's destination. Note that when using this transcoding, the stream will be trancoded to 30 FPS and CBR bit rate. The value is a JSON object with the support for same optoins as an entry in `transcodeProfiles`. Prefer to use `transcodeProfile` key if you have multiple outputs using the same encoding configuration.
  * `transcodeProfile` - If present, this stream will use the transcode settings provided by the `transcodeProfile` referenced by the string identifier set here. If multiple streams refer to the same `transcodeProfile`, the transcoding will only be done once. This option is mutually exclusive with the `transcode` key.


Here is an example of the JSON configuration file:
```json
{
  "endpoint": "live",
  "transcodeProfiles": {
    "720_60fps": {
      "pixels": "1280x720",
      "videoBitRate": "4500k",
      "videoFrameRate": 60
    }
  },
  "rebroadcastList": [
    {
      "name": "linkedin-via-restream",
      "platform": "restream",
      "streamKey": "whatever",
      "transcode": {
        "pixels": "1920x1080",
        "videoBitRate": "3500k",
        "videoKeyFrameSecs": 2,
        "audioBitRate": "160k",
        "audioSampleRate": 44100
      }
    },  
    {
      "name": "youtube",
      "platform": "youtube",
      "streamKey": "abc123"
    },
    {
      "name": "facebook-1",
      "platform": "facebook",
      "streamKey": "def456",
      "transcodeProfile": "720_60fps"
    },
    {
      "name": "facebook-2",
      "platform": "facebook",
      "streamKey": "ghi789",
      "transcodeProfile": "720_60fps"
    },
    {
      "name": "periscope",
      "platform": "periscope",
      "regionCode": "ca",
      "streamKey": "jkl012",
      "transcode": {
        "pixels": "1280x720",
        "videoBitRate": "3500k",
        "videoKeyFrameSecs": 3,
        "audioBitRate": "128k",
        "audioSampleRate": 44100
      }
    }
  ]
}
```
Note that as long as the `name` elements are different, you can have more than one destination pushing to the same `platform`, each with different `streamKey` values.

If you would like to capture a recording of the stream sent to the ingest endpoint, bind a local directory on your host to the `/var/www/html/recordings/` file path within the Docker image when launching the Docker container. You can also see statistics about the various streams this server is pushing by visiting this web page: `http://__docker_host_IP_address__/stat`. You can optionally disable recording by setting the env var `CONFIG_DISABLE_RECORD` to a non-empty value.

Once the Docker image is running, set up your stream software with the following parameters:

* **Server** : `rtmp://__docker_host_IP_address__/__endpoint_name__` - Replace `__docker_host_IP_address__` with the IP address of your host that is running this Docker container. Also replace `__endpoint_name__` with the value used for the `endpoint` element in your JSON configuration file.
* **Stream Key** : `__made_up_stream_name__?pwd=__made_up_password__` - Here `__made_up_stream_name__` is any arbitrary stream name, and `__made_up_password__` is the same password defined for `MULTISTREAMING_PASSWORD` above.

In OBS, you would set the above parameters for a "Custom..." Service in the Stream settings.

The next thing to do is start your stream!

## Development

In order to develop on this repository, you must have a python 3.7 environment and [`Pipenv`](https://pypi.org/project/pipenv/) installed.

- Clone the repository
- Navigate to the `multistreaming-server/` directory
- Install dependencies and create the virtual env by running `pipenv install`
- Activate the virtual env by `pipenv shell`

## Future work

Goals for future improvements to this project include:

* Adding more streaming services
* Creating a useful status web page
* Create a control panel web page where you'd set stream keys rather than through a JSON file

Pull requests are welcome!

## Acknowledgements

This work was heavily influenced by the following articles:

* The basic concept: [https://www.scaleway.com/en/docs/setup-rtmp-streaming-server/](https://www.scaleway.com/en/docs/setup-rtmp-streaming-server/)
* How to use `rtmps` with Facebook: [https://dev.to/lax/rtmps-relay-with-stunnel-12d3](https://dev.to/lax/rtmps-relay-with-stunnel-12d3)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]
* Bummped various package versions.
* Added python `requirements.txt` file for UpCloud deployment script.

## [0.4.2]
### Added
* Support for ReStream, and by extension, platforms that ReStream supports, such as LinkedIn.

### Changed
* Bumped jinja2 from 2.11.2 to 2.11.3
* Updated `apk` to use the `latest-stable` rather than `edge` as `edge` appears to not be updated.
* Updated `ffmpeg`, `ngtnx`, `superviserd`, and `pipenv` to their latest versions.

## [0.4.1]
### Changed
* Updated `ffmpeg` to the 4.3.x version branch

### Fixed
* Corrected bug that occurred when audio bit rated was configured in JSON using an integer
* Corrected a bug where each transcode block pushed to all destinations

## [0.4.0]
### Added
* Support for Mixcloud and DLive.
* Added support for specifying transcode profiles that can be used across multiple destination configurations.
* Add kubernetes manifests, fix nginx logging, streamline the dockerfile

### Changed
* Refactored how processes are managed within the Docker container. Migrated to `jinja2`, use `pipfile` and `supervisord`.

### Fixed
* Corrected a spelling error in the file path that the Docker image looks for the JSON configuration.
* Adjusted the timing of the UpCloud deployment script to better handle timing variation in the cloud service.
* Fixed a bug in the UpCloud deployment script that manifested when the configuration filepath has spaces in it.
* Changed `localhost` references to `127.0.0.1` to prevent IPv6 resolution. 

## [0.3.1]
### Added
* Created a deployment script for running the `multistreaming-server` locally.
* Created a deployment script for the [UpCloud cloud hosting service](https://upcloud.com/signup/?promo=A2CVWA).

### Changed
* Improved error handling in the RTMP configuration generation script.
* Improved error handling in the Lindode deploy script.

### Fixed
* Correct bug in RTMP configuration generation when the audio bit rate is set.
* Reinstated support for Microsoft Stream
* Addressed coding style issues with the python script `rtmp-conf-generator.py`

## [0.3.0]
### Changed
* **This is a major change that breaks any deployment based on prior version.**
* Migrated away from using environment variables to confihure rebroadcasting destinations and instead now used a JSON file.
* Cleanly enables rebroadcasting to multiple stream keys on the same platform.
* Migrated

### Added
* Created a deployment script for the [Linode cloud hosting service](https://www.linode.com/?r=37246e0d6a6198293308e698647804fbfe02845e).

### Fixed
* Corrected how transcoding is implemented: larger buffer, better audio

## [0.2.0]
### Changed
* Migrated Docker base to `jrottenberg/ffmpeg:4.2-alpine` in order to get a more robust `ffmpeg` compilation.
* Migrated RTMP module to use a more recent and currently maintained branch.

### Added
* Ability to transcode the stream being sent to Facebook.

## 0.1.2
* First verison!


[Unreleased]: https://github.com/michaelkamprath/multi-service-rtmp-broadcaster/compare/v0.4.2...HEAD
[0.4.2]: https://github.com/michaelkamprath/multi-service-rtmp-broadcaster/compare/v0.4.1...v0.4.2
[0.4.1]: https://github.com/michaelkamprath/multi-service-rtmp-broadcaster/compare/v0.4.0...v0.4.1
[0.4.0]: https://github.com/michaelkamprath/multi-service-rtmp-broadcaster/compare/v0.3.1...v0.4.0
[0.3.1]: https://github.com/michaelkamprath/multi-service-rtmp-broadcaster/compare/v0.3.0...v0.3.1
[0.3.0]: https://github.com/michaelkamprath/multi-service-rtmp-broadcaster/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/michaelkamprath/multi-service-rtmp-broadcaster/compare/v0.1.2...v0.2.0


deployment-scripts/local-deploy.sh
#!/bin/bash

print_help_and_exit()
{
  echo "Usage: "
  echo "    local-deploy.sh [-c /path/to/config.json] [-s <streaming password>] [-d <docker image name>] [-b]"
  echo ""
  echo "Environment Variables (can be used instead of command arguments): "
  echo "    RTMP_SERVER_CONFIG_FILEPATH    : path to RTMP server config JSON file. Can replace -c option."
  echo "    RTMP_SERVER_STREAM_PASSWORD    : Password used by RTMP streaming server for livestreams. Can replace -s option."
  echo "    RTMP_SERVER_DOCKER_IMAGE_NAME  : The image name to pull from Docker hub, or the image name to use if building locally (-b)."
  exit 0
}


build_docker_image_locally()
{
  # assumes this script is running in its original git repository.
  BASEDIR=$(dirname "$BASH_SOURCE")
  if [ $RTMP_SERVER_DOCKER_IMAGE_NAME == $DEFAULT_DOCKER_HUB_IMAGE_NAME ]; then
    RTMP_SERVER_DOCKER_IMAGE_NAME="multistreaming-server:latest"
  fi
  echo "Building Docker image '${RTMP_SERVER_DOCKER_IMAGE_NAME}'"
  docker build -t $RTMP_SERVER_DOCKER_IMAGE_NAME $BASEDIR/../multistreaming-server/
}

DEFAULT_DOCKER_HUB_IMAGE_NAME="kamprath/multistreaming-server:latest"
RTMP_SERVER_DOCKER_IMAGE_NAME=${RTMP_SERVER_DOCKER_IMAGE_NAME:-$DEFAULT_DOCKER_HUB_IMAGE_NAME}

build_locally=false
# check arguments passed to this script
while getopts "c:s:d:bh" option; do
  case "${option}" in
    b)
      # By default this script pulls from Docker hub. This ooption
      # forces a local build of the Docker image.
      build_locally=true
      ;;
    c)
      RTMP_SERVER_CONFIG_FILEPATH=$OPTARG
      ;;
    s)
      RTMP_SERVER_STREAM_PASSWORD=$OPTARG
      ;;
    d)
      RTMP_SERVER_DOCKER_IMAGE_NAME=$OPTARG
      ;;
    h)
      print_help_and_exit
      ;;
  esac
done
shift $((OPTIND -1))

if [ -z $RTMP_SERVER_STREAM_PASSWORD ]; then
  echo "You must define a streaming password for the streaming server. Use the '-s' option or set the RTMP_SERVER_STREAM_PASSWORD environment variable."
  exit 2
fi

if [ -z $RTMP_SERVER_CONFIG_FILEPATH ]; then
  echo "You must define a filepath to the streaming server configuration JSON file. Use the -c option."
fi

# Build image if needed
if [ "$build_locally" = true ]; then
    build_docker_image_locally
fi

# Launch the docker image
config_file_absolute_path="$(cd "$(dirname "$RTMP_SERVER_CONFIG_FILEPATH")"; pwd)/$(basename "$RTMP_SERVER_CONFIG_FILEPATH")"
echo "Launching $RTMP_SERVER_DOCKER_IMAGE_NAME Docker image with config at '${config_file_absolute_path}'."
docker_proc_id=$( \
  docker run -d -p 80:80 -p 1935:1935 \
    --env MULTISTREAMING_PASSWORD=${RTMP_SERVER_STREAM_PASSWORD} \
    -v "${config_file_absolute_path}":/rtmp-configuration.json \
    ${RTMP_SERVER_DOCKER_IMAGE_NAME} \
)
docker_short_proc_id=$(echo $docker_proc_id | cut -c1-12)
echo ""
echo "Launched docker container $docker_short_proc_id. To stop this contianer:"
echo "    docker stop $docker_short_proc_id"
echo ""
echo "Visit the Multistreaming Server's statistics page here:"
echo "    http://127.0.0.1/stat"
echo ""


deployment-scripts/linode-deploy.sh
#!/bin/bash

#
# Assumes that the Linode CLI tool is already installed and enabled to create linodes.
#

# helper functions
print_help_and_exit()
{
  echo "Usage: "
  echo "    linode-deploy.sh [-c /path/to/config.json] [-p <new server root password>] [-k /path/to/ssh_key.pub] [-s <streaming password>]"
  echo ""
  echo "Environment Variables (can be used instead of command arguments): "
  echo "    RTMP_SERVER_CONFIG_FILEPATH    : path to RTMP server config JSON file. Can replace -c option."
  echo "    RTMP_SERVER_ROOTPW             : Root password to be set for newly created Linode server. Can replace -p option."
  echo "    RTMP_SERVER_AUTORIZED_KEY_FILE : SSH public key file path to enable secure connections for newly created Linode server. Defaults to  ~/.ssh/id_rsa.pub. Can replace -k option."
  echo "    RTMP_SERVER_STREAM_PASSWORD    : Password used by RTMP streaming server for livestreams. Can replace -s option."
  echo "    RTMP_SERVER_LINODE_REGION      : Linode region to create server in. Defaults to 'us-west'."
  echo "    RTMP_SERVER_LINODE_NODE_TYPE   : Linode node type to use for server. Defaults to 'g6-standard-4'."
  echo ""
  exit 0
}

# check arguments passed to this script
while getopts "c:p:k:s:h" option; do
  case "${option}" in
  c)
    RTMP_SERVER_CONFIG_FILEPATH=$OPTARG
    ;;
  p)
    RTMP_SERVER_ROOTPW=$OPTARG
    ;;
  k)
    RTMP_SERVER_AUTORIZED_KEY_FILE=$OPTARG
    ;;
  s)
    RTMP_SERVER_STREAM_PASSWORD=$OPTARG
    ;;
  h)
    print_help_and_exit
    ;;
  esac
done
shift $((OPTIND -1))

if [ -z $RTMP_SERVER_STREAM_PASSWORD ]; then
  echo "You must define a streaming password for the streaming server. Use the '-s' option or set the RTMP_SERVER_STREAM_PASSWORD environment variable."
  exit 2
fi

if [ -z $RTMP_SERVER_ROOTPW ]; then
  echo "You must define a root password for the server. Use the '-p' option or set the RTMP_SERVER_ROOTPW environement variable."
  exit 2
fi

if [ -z $RTMP_SERVER_CONFIG_FILEPATH ]; then
  echo "You must define a filepath to the streaming server configuration JSON file. Use the -c option."
fi

# Set variables to defaults if unset
RTMP_SERVER_LINODE_REGION=${RTMP_SERVER_LINODE_REGION:-"us-west"}
RTMP_SERVER_LINODE_NODE_TYPE=${RTMP_SERVER_LINODE_NODE_TYPE:-"g6-standard-4"}
RTMP_SERVER_AUTORIZED_KEY_FILE=${RTMP_SERVER_AUTORIZED_KEY_FILE:-"~/.ssh/id_rsa.pub"}

LINODE_DOCKER_SCRIPT_ID="607433"

# launch a Docker server
echo "Creating Linode server ..."
linode_id=$( \
  linode-cli linodes create \
    --image "linode/debian9" \
    --label multistreaming-server \
    --authorized_keys "$(cat $RTMP_SERVER_AUTORIZED_KEY_FILE)" \
    --root_pass "$RTMP_SERVER_ROOTPW" \
    --booted true \
    --region "$RTMP_SERVER_LINODE_REGION" \
    --type "$RTMP_SERVER_LINODE_NODE_TYPE" \
    --stackscript_id "$LINODE_DOCKER_SCRIPT_ID" \
    --text --format 'id' --no-headers \
)
lindode_status=$(linode-cli linodes view ${linode_id} --text --format 'status' --no-headers)
echo "Created Linode ${linode_id}. Current status is '${lindode_status}'. Waiting for it to transition to a running state."
while [ $lindode_status != "running" ]
do
  sleep 10
  lindode_status=$(linode-cli linodes view ${linode_id} --text --format 'status' --no-headers)
  echo "Linode ${linode_id} status is '${lindode_status}'."
done
linode_ip_addr=$(linode-cli linodes view ${linode_id} --text --format 'ipv4' --no-headers)
echo "Linode ${linode_id} creation complete. IPv4 address is: ${linode_ip_addr}"
echo "Waiting 5 minutes for booting and Docker installation to complete."
sleep 60
echo "Waiting 4 minutes for booting and Docker installation to complete."
sleep 60
echo "Waiting 3 minutes for booting and Docker installation to complete."
sleep 60
echo "Waiting 2 minutes for booting and Docker installation to complete."
sleep 60
echo "Waiting 1 minutes for booting and Docker installation to complete."
sleep 60


# Add IP address to known hosts so ssh commands can proceed unimpeded
echo "Addding ${linode_ip_addr} to known_hosts"
ssh-keygen -R ${linode_ip_addr}
ssh-keyscan -T 240 ${linode_ip_addr} >> ~/.ssh/known_hosts

# copy the passed RTMP server config file to the Server
echo "Copying '${RTMP_SERVER_CONFIG_FILEPATH}' to path /root/rtmp_server_config.json on Linode ${linode_id}"
scp -q "$RTMP_SERVER_CONFIG_FILEPATH" root@"${linode_ip_addr}":/root/rtmp_server_config.json
if [ $? -ne 0 ]; then
  echo "ERROR - Unable to copy '${RTMP_SERVER_CONFIG_FILEPATH}' to linode ${linode_id}. Please terminate the Linode with this command:"
  echo "    linode-cli linodes delete ${linode_id}"
  exit 1
fi
# now launch docker container
echo "Launch kamprath/multistreaming-server:latest Docker image."
docker_proc_id=$(ssh root@"$linode_ip_addr" \
  "docker run -d -p 80:80 -p 1935:1935 " \
    "--env MULTISTREAMING_PASSWORD=${RTMP_SERVER_STREAM_PASSWORD} " \
    "-v /root/rtmp_server_config.json:/rtmp-configuration.json " \
    "kamprath/multistreaming-server:latest" \
)
echo "Started Docker container: $docker_proc_id"
# Finished
echo "Finished!"
echo ""
echo "The IP address for the Multistreaming Server is:"
echo "    ${linode_ip_addr}"
echo ""
echo "Visit the Multistreaming Server's statistics page here:"
echo "    http://${linode_ip_addr}/stat"
echo ""
echo "Use this command to kill the server:"
echo "    linode-cli linodes delete ${linode_id}"
echo ""
echo "Use this command to log into the server:"
echo "    ssh root@${linode_ip_addr}"
echo ""


multistreaming-server/rtmp-conf-generator.py
#!/usr/bin/env python3

import copy
import jinja2
import json
import os
import re
import sys

#
# Configurable ENV VARS
#

CONFIG_NGINX_DEBUG = os.getenv('CONFIG_NGINX_DEBUG', False)
CONFIG_FFMPEG_LOG = os.getenv('CONFIG_FFMPEG_LOG', False)
CONFIG_FFMPEG_MAX_MUXING_QUEUE_SIZE = os.getenv(
    'CONFIG_FFMPEG_MAX_MUXING_QUEUE_SIZE', False
)
CONFIG_DISABLE_RECORD = os.getenv('CONFIG_DISABLE_RECORD', False)

#
# Defaults
#

RTMP_TRANSCODE_AUDIO_OPTS_COPY = '-c:a copy'
RTMP_TRANSCODE_AUDIO_OPTS_CUSTOM = (
    '-c:a libfdk_aac -b:a %%AUDIO_BIT_RATE%% -ar %%AUDIO_SAMPLE_RATE%%'
)

# https://support.restream.io/en/articles/73108-best-settings
PUSH_URL_RESTREAM = 'rtmp://live.restream.io/live/%%STREAM_KEY%%' # Restream let's you stream to LinkedIn Live
PUSH_URL_YOUTUBE = 'rtmp://127.0.0.1:19352/live2/%%STREAM_KEY%%'
PUSH_URL_FACEBOOK = 'rtmp://127.0.0.1:19350/rtmp/%%STREAM_KEY%%'
PUSH_URL_TWITCH = 'rtmp://live-cdg.twitch.tv/app/%%STREAM_KEY%%'
PUSH_URL_INSTAGRAM = 'rtmp://127.0.0.1:19351/rtmp/%%STREAM_KEY%%'
PUSH_URL_PERISCOPE = 'rtmp://%%REGION_CODE%%.pscp.tv:80/x/%%STREAM_KEY%%'
PUSH_URL_MICROSOFT_STREAM = '%%RTMP_URL%% app=live/%%APP_NAME%%'
PUSH_URL_MIXCLOUD = 'rtmp://rtmp.mixcloud.com/broadcast/%%STREAM_KEY%%'
PUSH_URL_DLIVE = 'rtmp://stream.dlive.tv/live/%%STREAM_KEY%%'


DEFAULT_TRANSCODE_CONFIG = {
    'pixels': '1280x720',
    'videoBitRate': '4500k',
    'videoFrameRate': 60,
    'keyFrames': 60,
    'audioOpts': '-c:a copy',
    'logFfmpeg': True if CONFIG_FFMPEG_LOG else False,
    'maxMuxingQueueSize': int(CONFIG_FFMPEG_MAX_MUXING_QUEUE_SIZE)
    if CONFIG_FFMPEG_MAX_MUXING_QUEUE_SIZE
    else None,
}

DEFAULT_AUDIO_OPTS = {
    'audioBitRate': '160k',
    'audioSampleRate': '48000',
}


def generatePlatormPushURL(block_config):
    if 'platform' not in block_config:
        print('ERROR - Application block is missing platform element.', file=sys.stderr)
        exit(1)
    push_url = 'push-it-real-good'
    
    if block_config['platform'] == 'restream':
        push_url = PUSH_URL_RESTREAM.replace('%%STREAM_KEY%%', block_config['streamKey'])
    elif block_config['platform'] == 'youtube':
        push_url = PUSH_URL_YOUTUBE.replace('%%STREAM_KEY%%', block_config['streamKey'])
    elif block_config['platform'] == 'facebook':
        # must push through stunnel. Push through Facebook stunnel port.
        push_url = PUSH_URL_FACEBOOK.replace(
            '%%STREAM_KEY%%', block_config['streamKey']
        )
    elif block_config['platform'] == 'twitch':
        push_url = PUSH_URL_TWITCH.replace('%%STREAM_KEY%%', block_config['streamKey'])
    elif block_config['platform'] == 'instagram':
        # must push through stunnel. Push through Instagram stunnel port.
        push_url = PUSH_URL_INSTAGRAM.replace(
            '%%STREAM_KEY%%', block_config['streamKey']
        )
    elif block_config['platform'] == 'periscope':
        region_code = (
            block_config['regionCode'] if 'regionCode' in block_config else 'ca'
        )
        push_url = PUSH_URL_PERISCOPE.replace(
            '%%STREAM_KEY%%', block_config['streamKey']
        ).replace('%%REGION_CODE%%', region_code)
    elif block_config['platform'] == 'custom':
        push_url = block_config['customRTMPURL']
    elif block_config['platform'] == 'microsoft-stream':
        ms_source_url = block_config['fullRTMPURL']
        ms_rtmp_url = re.search(r'^(.*)/live/', ms_source_url).group(1)
        ms_app_name = re.search(r'/live/(.*)$', ms_source_url).group(1)
        push_url = PUSH_URL_MICROSOFT_STREAM.replace(
            '%%RTMP_URL%%', ms_rtmp_url
        ).replace('%%APP_NAME%%', ms_app_name)
    elif block_config['platform'] == 'mixcloud':
        push_url = PUSH_URL_MIXCLOUD.replace(
            '%%STREAM_KEY%%', block_config['streamKey']
        )
    elif block_config['platform'] == 'dlive':
        push_url = PUSH_URL_DLIVE.replace('%%STREAM_KEY%%', block_config['streamKey'])
    else:
        print(
            'ERROR - an unsupported platform type was provided in destination configation',
            file=sys.stderr,
        )
        exit(1)
    return push_url


def generateTranscodeConfig(transcode_config_name, block_config, config):
    block_config_name = block_config['name']
    default_transcode_config = DEFAULT_TRANSCODE_CONFIG.copy()
    transcode_config_block = config.get('transcodeProfiles', {}).get(
        transcode_config_name, block_config.get('transcode')
    )
    if not transcode_config_block:
        print(
            f'ERROR - unable to resolve transcode profile for {transcode_config_name} in block {block_config_name}',
            file=sys.stderr,
        )
        exit(1)
    transcode_config = {
        key: copy.deepcopy(transcode_config_block.get(key, DEFAULT_TRANSCODE_CONFIG[key]))
        for key in default_transcode_config.keys()
    }
    transcode_config['applicationEndpoints'] = set()

    if 'videoKeyFrameSecs' in transcode_config_block:
        transcode_config['keyFrames'] = 30 * transcode_config_block['videoKeyFrameSecs']

    if ('audioBitRate' in transcode_config_block) or (
        'audioSampleRate' in transcode_config_block
    ):
        transcode_config['audioOpts'] = RTMP_TRANSCODE_AUDIO_OPTS_CUSTOM.replace(
            '%%AUDIO_BIT_RATE%%',
            str(transcode_config_block.get(
                'audioBitRate', DEFAULT_AUDIO_OPTS['audioBitRate']
            )),
        ).replace(
            '%%AUDIO_SAMPLE_RATE%%',
            str(transcode_config_block.get(
                'audioSampleRate', DEFAULT_AUDIO_OPTS['audioSampleRate']
            )),
        )
    return transcode_config


def loadJsonConfig(path):
    try:
        with open(path, 'r') as f:
            config = json.load(f)
    except json.decoder.JSONDecodeError as err:
        print(
            "ERROR decoding JSON config file '{0}': {1}".format(path, err),
            file=sys.stderr,
        )
        exit(1)
    except:
        print("ERROR loading JSON config file '{0}'".format(path), file=sys.stderr)
        exit(1)

    return config


def generateConfig(config_file, nginx_config_template):
    config = loadJsonConfig(config_file)
    # unfortunate hack to make code compatible with previous spelling error
    config_list_key = (
        'rebroacastList' if 'rebroacastList' in config else 'rebroadcastList'
    )

    record_mode = 'off' if CONFIG_DISABLE_RECORD else 'all'
    nginx_error_log = True if CONFIG_NGINX_DEBUG else False

    endpoint_name = config['endpoint'] if 'endpoint' in config else 'live'

    application_configs = {}
    transcode_configs = {}
    push_only_applications = set()

    for block_config in config[config_list_key]:
        if block_config.get('disabled', False):
            continue

        block_config_name = block_config['name']
        application_configs[block_config_name] = {
            'pushUrl': generatePlatormPushURL(block_config)
        }

        if ('transcodeProfile' in block_config) or 'transcode' in block_config:
            transcode_config_name = block_config.get(
                'transcodeProfile', f'inline_{block_config_name}'
            )

            if transcode_config_name not in transcode_configs:
                transcode_configs[transcode_config_name] = generateTranscodeConfig(
                    transcode_config_name, block_config, config
                )
            transcode_configs[transcode_config_name]['applicationEndpoints'].add(
                block_config_name
            )

        else:
            push_only_applications.add(block_config_name)

    with open(nginx_config_template) as fh:
        template = jinja2.Template(fh.read())

    return template.render(
        nginx_error_log=nginx_error_log,
        endpoint_name=endpoint_name,
        record_mode=record_mode,
        transcode_configs=transcode_configs,
        push_only_applications=push_only_applications,
        application_configs=application_configs,
    )


if __name__ == '__main__':
    if len(sys.argv) != 3:
        print(
            'Must pass two arguments of the JSON configuration file path and the nginx config.',
            file=sys.stderr,
        )
        sys.exit(1)
    rtmp_conf = generateConfig(sys.argv[1], sys.argv[2])
    print(rtmp_conf)
